### PR TITLE
feat(mlx): add handler for aten.roll

### LINKED ISSUE #18919: Good First Issue: Add MLX Op Handler for `aten.roll`
# Good First Issue: Add MLX Op Handler for `aten.roll`

## Summary

Add support for `aten.roll` in the MLX delegate. This op shifts tensor elements along specified dimensions with wrap-around and is needed by Swin Transformer's shift-window attention mechanism.

## Background

The MLX delegate converts PyTorch aten ops into MLX graph nodes during export. While `aten.roll` decomposes into `index_select` + `arange` + `cat` operations, a native MLX implementation using `mlx::core::roll` would be more efficient (single kernel vs multiple ops).

### Approach: New schema node + runtime

This requires a new `RollNode` because MLX has a dedicated `roll` function that's more efficient than the decomposed representation.

## Steps

1. **Add node to `backends/mlx/serialization/schema.fbs`**
   ```flatbuffers
   table RollNode {
     x: Tid;
     out: Tid;
     shift: [IntOrVid];  // shift amounts per axis
     axes: [int];        // axes to roll
   }
   ```
   Add `RollNode` to the `OpNode` union (append only, do not reorder).

2. **Regenerate serialization code**
   ```bash
   python backends/mlx/serialization/generate.py
   ```

3. **Add C++ runtime exec function in `backends/mlx/runtime/MLXInterpreter.h`**
   ```cpp
   void exec_RollNode(const RollNode& node) {
     auto x = get_tensor(node.x());
     
     std::vector<int> shifts;
     for (auto s : *node.shift()) {
       shifts.push_back(resolve_int_or_vid(s));
     }
     
     std::vector<int> axes;
     for (auto a : *node.axes()) {
       axes.push_back(a);
     }
     
     auto out = mlx::core::roll(x, shifts, axes, stream_);
     set_tensor(node.out(), out);
   }
   ```

4. **Add handler in `backends/mlx/ops.py`**
   
   Since `RollNode` is a simple unary-style op (input tensor → output tensor), add it to the `_UNARY_OPS` table:

   ```python
   # In the _UNARY_OPS list, add:
   (torch.ops.aten.roll.default, RollNode, "aten.roll"),
   ```
   
   **Note:** This only works if `RollNode` follows the standard unary signature `(x, out)`. However, `roll` has additional parameters (`shifts`, `axes`), so you may need a custom handler instead:

   ```python
   @REGISTRY.register(target=[torch.ops.aten.roll.default])
   def _roll_handler(P: MLXProgramBuilder, n: Node) -> Slot:
       args = P.args(n)
       require_args(args, 2, 3, "aten.roll")
       require_kwargs(P.kwargs(n), set(), "aten.roll")
       x = args[0]
       shifts = args[1]
       dims = args[2] if len(args) > 2 else []
       
       # Normalize shifts and dims to lists
       if isinstance(shifts, int):
           shifts = [shifts]
       if isinstance(dims, int):
           dims = [dims]
       if not dims:
           dims = list(range(len(n.args[0].meta["val"].shape)))
       
       out = P.make_or_get_slot(n)
       P.emit(
           RollNode(
               x=P.slot_to_tid(x),
               out=P.slot_to_tid(out),
               shift=[P.to_int_or_vid(s) for s in shifts],
               axes=list(dims),
           )
       )
       return out
   ```

5. **Add test in `backends/mlx/test/test_ops.py`**
   
   This op doesn't fit the simple unary pattern, so create a custom test class:

   ```python
   class RollModel(nn.Module):
       def __init__(self, shifts: int, dims: int):
           super().__init__()
           self.shifts = shifts
           self.dims = dims
       
       def forward(self, x: torch.Tensor) -> torch.Tensor:
           return torch.roll(x, shifts=self.shifts, dims=self.dims)

   @register_test
   class RollTest(OpTestCase):
       name = "roll"
       
       def __init__(self, shape: Tuple[int, ...], shifts: int, dims: int):
           self.shape = shape
           self.shifts = shifts
           self.dims = dims
           self.name = f"roll_shift{shifts}_dim{dims}"
       
       @classmethod
       def get_test_configs(cls) -> List["RollTest"]:
           return [
               cls(shape=(8,), shifts=2, dims=0),
               cls(shape=(4, 5), shifts=1, dims=0),
               cls(shape=(4, 5), shifts=-2, dims=1),
               cls(shape=(3, 4, 5), shifts=3, dims=2),
           ]
       
       def create_model(self) -> nn.Module:
           return RollModel(self.shifts, self.dims)
       
       def create_inputs(self) -> Tuple[torch.Tensor, ...]:
           return (torch.randn(self.shape),)
   ```

### Running tests

```bash
python -m executorch.backends.mlx.test.run_all_tests -k roll
```

## References

- MLX C++: `array roll(const array &a, int shift, int axis, StreamOrDevice s = {})`
- PyTorch signature: `roll(Tensor self, int[1] shifts, int[1] dims=[]) -> Tensor`
- Use case: Swin Transformer shift-window attention

### PR DESCRIPTION
**Pull request by ishangodawatta**
## Summary

Adds an MLX delegate handler for `aten.roll`, mapping `torch.roll` onto `mlx::core::roll` via a new `RollNode` in the schema. Replaces the default decomposition (`index_select + arange + cat`) with a single native kernel — needed by Swin Transformer's shift-window attention.

Flat roll (`dims=[]`) raises `NotImplementedError` for now; no known consumer needs it yet.

Generated files (`MLXLoader.*`, `schema_generated.h`, `mlx_graph_schema.py`, `_generated_serializers.py`, `_generated_inspector.py`, `_generated/`) are regenerated from `schema.fbs` by `backends/mlx/CMakeLists.txt` at build time and are deliberately not committed.

Fixes #18919.

## Test plan

- `python backends/mlx/serialization/generate.py` — regenerates cleanly with `RollNode` in all expected outputs.
- `lintrunner --skip MYPY --paths-cmd 'git diff --name-only upstream/main'` — no issues.
- End-to-end `run_all_tests -k roll` not run locally (no executorch build on this machine); relying on CI. Happy to push fixes if it finds anything.

cc @metascroy

### CODE REVIEW

**Comment by metascroy** [comment]
Looks fantastic @IshanG97!

Let's see what CI says, but is there a reason you didn't run the test locally? Ideally, I'd like to see the output of the test in the PR summary :)

**Review by metascroy — APPROVED** [review]

**Comment by metascroy** [comment]
Ignore all of the test-mlx-llm result failures in CI, they have to do with an HF token issue on external contributions.

**Comment by ishangodawatta** [comment]
> Looks fantastic @IshanG97!
> 
> Let's see what CI says, but is there a reason you didn't run the test locally? Ideally, I'd like to see the output of the test in the PR summary :)

Hi @metascroy! Sorry for the late response, I didn't initially because I couldn't get the latest version of Xcode on my machine, I had to use a proxy (to be honest, I was relying on the CI to catch anything haha!). I've now downloaded Xcode and run the tests locally, see my trimmed output:

```
  $ python -m executorch.backends.mlx.test.run_all_tests roll -v

  ============================================================
  Running test: roll_shift(2)_dim(0)
  ============================================================

  Step 1: Generating test files...
  Exporting model to .../op_tests/roll_shift(2)_dim(0)/model.pte

  ============================================================
  EXPORTED PROGRAM (torch.export)
  ============================================================
  ExportedProgram:
      class GraphModule(torch.nn.Module):
          def forward(self, x: "f32[8]"):
              roll: "f32[8]" = torch.ops.aten.roll.default(x, [2], [0]);  x = None
              return (roll,)

  ============================================================
  EDGE PROGRAM (after decomposition)
  ============================================================
  ExportedProgram:
      class GraphModule(torch.nn.Module):
          def forward(self, x: "f32[8]"):
              lowered_module_0 = self.lowered_module_0
              executorch_call_delegate = torch.ops.higher_order.executorch_call_delegate(lowered_module_0, x);  lowered_module_0 = x = None
              getitem: "f32[8]" = executorch_call_delegate[0];  executorch_call_delegate = None
              return (getitem,)

  ======================================================================
  MLX Graph Summary
  ======================================================================
  Version: 1
  Constant tensors: 0
  Input tensors: 1
  Output tensors: 1
  Instruction chains: 1
  Main chain idx: 0

  Chain 0 (main) (1 instructions):
    [0] RollNode
        x: tid 0
        out: tid 1
        shift: [{'literal': 2}]
        axes: [0]

  Inputs:  tid 0
  Outputs: tid 1
  ======================================================================

  Step 2: Verifying MLX delegation...
    Expected MLX segments: 1
    Actual MLX segments:   1
  ✓ MLX delegation verified

  Step 3: Running C++ binary...
  C++ binary output: Loading model...
  Model loaded successfully
    Input 0: dtype=0, shape=[8]
  Executing forward...
  Execution succeeded, 1 outputs
    Output 0: dtype=0, shape=[8]
  OK

  Step 4: Comparing outputs...
  ✓ PASSED: All outputs match

  ============================================================
  Running test: roll_shift(1)_dim(0)        ... ✓ PASSED
  Running test: roll_shift(-2)_dim(1)       ... ✓ PASSED
  Running test: roll_shift(3)_dim(2)        ... ✓ PASSED
  Running test: roll_shift(1,2)_dim(0,2)    ... ✓ PASSED
  Running test: roll_shift(-1,-2,-3)_dim(0,1,2) ... ✓ PASSED
  Running test: roll_shift(2)_dim(-1)       ... ✓ PASSED
  ============================================================

  ============================================================
  TEST SUMMARY
  ============================================================
  Passed: 7
  Failed: 0
  ============================================================
```

**Review by metascroy — APPROVED** [review]
Looks good, running CI again

**Review by metascroy — APPROVED** [review]

**Comment by metascroy** [comment]
Thanks for the contribution @IshanG97!
